plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryArrowToPageConverter.java

```diff
@@ -0,0 +1,294 @@
+            }
+        }
+        catch (ClassCastException ex) {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Unhandled type for %s: %s", javaType.getSimpleName(), type), ex);
+        }
+    }
+
+    private void writeVectorValues(BlockBuilder output, FieldVector vector, Consumer<Integer> consumer, int offset, int length)
+    {
+        for (int i = offset; i < offset + length; i++) {
+            if (vector.isNull(i)) {
+                output.appendNull();
+            }
+            else {
+                consumer.accept(i);
+            }
+        }
+    }
+
+    private void writeSlice(BlockBuilder output, Type type, FieldVector vector, int index)
+    {
+        if (type instanceof VarcharType) {
+            byte[] slice = ((VarCharVector) vector).get(index);
+            type.writeSlice(output, wrappedBuffer(slice));
+        }
+        else if (type instanceof VarbinaryType) {
+            byte[] slice = ((VarBinaryVector) vector).get(index);
+            type.writeSlice(output, wrappedBuffer(slice));
+        }
+        else {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, "Unhandled type for Slice: " + type.getTypeSignature());
+        }
+    }
+
+    private void writeObjectLongDecimal(BlockBuilder output, Type type, FieldVector vector, int index)
+    {
+        if (type instanceof DecimalType decimalType) {
+            verify(!decimalType.isShort(), "The type should be long decimal");
+            BigDecimal decimal = vector.getMinorType() == DECIMAL256 ? ((Decimal256Vector) vector).getObject(index) : ((DecimalVector) vector).getObject(index);
+            type.writeObject(output, Decimals.encodeScaledValue(decimal, decimalType.getScale()));
+        }
+        else {
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, "Unhandled type for Object: " + type.getTypeSignature());
+        }
+    }
+
+    private void writeObjectShortDecimal(BlockBuilder output, DecimalType decimalType, FieldVector vector, int index)
+    {
+        verify(decimalType.isShort(), "The type should be short decimal");
+        BigDecimal decimal = vector.getMinorType() == DECIMAL256 ? ((Decimal256Vector) vector).getObject(index) : ((DecimalVector) vector).getObject(index);
+        decimalType.writeLong(output, encodeShortScaledValue(decimal, decimalType.getScale()));
+    }
+
+    private void writeObjectTimestampWithTimezone(BlockBuilder output, Type type, FieldVector vector, int index)
+    {
+        verify(type.equals(TIMESTAMP_TZ_MICROS));
+        long epochMicros = ((TimeStampVector) vector).get(index);
+        int picosOfMillis = toIntExact(floorMod(epochMicros, MICROSECONDS_PER_MILLISECOND)) * PICOSECONDS_PER_MICROSECOND;
+        type.writeObject(output, fromEpochMillisAndFraction(floorDiv(epochMicros, MICROSECONDS_PER_MILLISECOND), picosOfMillis, UTC_KEY));
```

Comment:
It would be nice to leave a comment why specifying `UTC_KEY` is safe because the method is for timestamp with time zone type.

Reply:
Not sure how this comment should looks like